[Feature] 회원가입 API

jusang3057 opened the pull request:
**개요**
- 회원가입 API

**작업 내용**
- https://www.notion.so/25-03-30-API-1c69ae650b5980cc8324dfc77b59cc01


**수정**
1. request에 not-null적용

2. 버전구분 : [fun]2 -> [fun]V2

UserRepository.kt
   verifyCertificationCode: return token:String

CheckResponse.kt
   name : CheckResponse -> Duplicated
   data : BAD_REQUEST -> INVALID
   class : enum class -> sealed class

UserRepositoryImpl.kt
   postGeneralRegister(name, pho...) -> postGeneralRegister(token, nam...)
   postStudentRegister(name, pho...) -> postGeneralRegister(token, nam...)

jusang3057 committed:
[Feature] 회원가입 API

개요
* 회원가입 API
작업 내용
* https://www.notion.so/25-03-30-API-1c69ae650b5980cc8324dfc77b59cc01

kongwoojin commented on data/src/main/java/in/koreatech/koin/data/request/user/GeneralInfoRequest.kt:
```diff
@@ -0,0 +1,20 @@
+
+data class GeneralInfoRequest(
+    @SerializedName("name")
+    val name: String?,
```
request에는 필수값이라 not-null이 맞을 것 같습니다

kongwoojin commented on data/src/main/java/in/koreatech/koin/data/api/UserApi.kt:
```diff
@@ -50,4 +52,24 @@ interface UserApi {
+        @Query("phone") phone: String
+    )
+
+    @GET(URLConstant.USER.CHECKNICKNAME2)
```
그냥 2보단 V2가 나을 것 같습니다

kongwoojin commented on domain/src/main/java/in/koreatech/koin/domain/model/user/CheckResponse.kt:
```diff
@@ -0,0 +1,8 @@
+package `in`.koreatech.koin.domain.model.user
+
+enum class CheckResponse {
```
CheckResponse보단 조금 더 명확한 이름이면 좋을 것 같습니다

ThirFir replied:
도메인 레이어에서는 일반적으로 Response라는 네이밍으로 사용되진 않습니다

kongwoojin commented on domain/src/main/java/in/koreatech/koin/domain/model/user/CheckResponse.kt:
```diff
@@ -0,0 +1,8 @@
+
+enum class CheckResponse {
+    OK,
+    BAD_REQUEST,
```
그냥 BAD_REQUEST라고 반환하기 보단 다른 이름이 좋을 것 같습니다
API에서 bad request로 반환해주면 형식이 잘못됐다는거니깐 그런걸 반영하면 좋을 것 같습니다

kongwoojin commented on domain/src/main/java/in/koreatech/koin/domain/model/user/CheckResponse.kt:
```diff
@@ -0,0 +1,8 @@
+package `in`.koreatech.koin.domain.model.user
+
+enum class CheckResponse {
```
sealed class를 써도 괜찮을 것 같네요

kongwoojin commented on data/src/main/java/in/koreatech/koin/data/request/user/SmsVerifyRequest.kt:
```diff
@@ -0,0 +1,10 @@
+
+data class SmsVerifyRequest(
+    @SerializedName("phone_number")
+    val phoneNumber: String?,
```
여기도 마찬가지입니다

kongwoojin commented on data/src/main/java/in/koreatech/koin/data/request/user/StudentInfoRequest2.kt:
```diff
@@ -0,0 +1,24 @@
+
+data class StudentInfoRequest2(
+    @SerializedName("name")
+    val name: String?,
```
여기도요

kongwoojin commented on data/src/main/java/in/koreatech/koin/data/repository/UserRepositoryImpl.kt:
```diff
@@ -167,4 +171,135 @@ class UserRepositoryImpl @Inject constructor(
+        }
+    }
+
+    override suspend fun sendRegisterEmail(phoneNumber: String, certificationCode: String): Boolean {
```
email이 아니라 sms가 맞을 것 같습니다
그리고 전송이 아니라 인증번호 검증이니만큼 이걸 잘 나타낼 함수 이름이 좋을 것 같습니다

kongwoojin commented on data/src/main/java/in/koreatech/koin/data/api/UserApi.kt:
```diff
@@ -50,4 +52,24 @@ interface UserApi {
     ): Response<RefreshResponse>
+
+    @GET(URLConstant.USER.CHECKPHONE)
+    suspend fun checkPhone(
```
checkPhoneNumberDuplicate와 같이 무슨 API인지 보다 명확하게 이름을 짓는게 좋을 것 같습니다

kongwoojin commented on data/src/main/java/in/koreatech/koin/data/api/auth/UserAuthApi.kt:
```diff
@@ -123,4 +125,14 @@ interface UserAuthApi {
     )
+
+    @POST(URLConstant.USER.SMSSEND)
+    suspend fun smsVerify(
```
인증번호 전송 요청 및 인증 번호 검증 API에는 JWT 토큰이 필요하지 않습니다.
위치가 잘못됐습니다

kongwoojin commented on data/src/main/java/in/koreatech/koin/data/api/auth/UserAuthApi.kt:
```diff
@@ -123,4 +125,14 @@ interface UserAuthApi {
+    @POST(URLConstant.USER.SMSVERIFY)
+    suspend fun smsVerify(
+        @Body smsVerifyRequest: SmsVerifyRequest
+    )
```
인증번호 검증 이후에 임시 토큰을 반환해주는데 응답값이 없는 것 같습니다.

kongwoojin commented on data/src/main/java/in/koreatech/koin/data/api/UserApi.kt:
```diff
@@ -50,4 +52,24 @@ interface UserApi {
+    )
+
+    @POST(URLConstant.USER.STUDENT.REGISTER2)
+    suspend fun postStudentRegister(
```
회원가입 시 인증번호 검증 단계에서 받은 임시 토큰을 헤더로 넘겨줍니다.
헤더가 누락됐습니다

kongwoojin commented on domain/src/main/java/in/koreatech/koin/domain/repository/UserRepository.kt:
```diff
@@ -46,4 +47,34 @@ interface UserRepository {
+
+    suspend fun isPhoneDuplicated(phone: String): CheckResponse
+
+    suspend fun isStudentRegister(
```
자칫하면 회원가입이 됐는지 확인하는 API나 다른걸로 오해할 소지가 있어 보입니다.
이 함수에서 호출하는 함수와 마찬가지로 postStudentRegister 정도의 이름이 적당해보입니다

kongwoojin commented on domain/src/main/java/in/koreatech/koin/domain/repository/UserRepository.kt:
```diff
@@ -46,4 +47,34 @@ interface UserRepository {
+        nickname: String
+    ): Boolean
+
+    suspend fun isGeneralRegister(
```
마찬가지입니다

kongwoojin commented on domain/src/main/java/in/koreatech/koin/domain/error/user/UserErrorHandler.kt:
```diff
@@ -18,4 +18,6 @@ interface UserErrorHandler {
 
     fun handleVerifyUserPasswordError(throwable: Throwable): ErrorHandler
+
+    fun handlePhoneDuplicatedError(throwable: Throwable): ErrorHandler
```
사용하지 않는 함수 같습니다

kongwoojin reviewed (changes requested):
고생하셨습니다!
코멘트 확인 해주세요

추가로 lint 수정도 부탁드립니다

ThirFir commented on data/src/main/java/in/koreatech/koin/data/repository/UserRepositoryImpl.kt:
```diff
@@ -167,4 +171,135 @@ class UserRepositoryImpl @Inject constructor(
+            userRemoteDataSource.checkNickname2(nickname)
+            CheckResponse.OK
+        } catch (e: HttpException) {
+            if (e.code() == 200) {
+                CheckResponse.OK
+            } else if (e.code() == 400) {
+                CheckResponse.BAD_REQUEST
+            } else if (e.code() == 409) {
+                CheckResponse.CONFLICT
+            } else {
+                CheckResponse.UNDEFINED
+            }
```
when문을 사용해봐도 좋을 듯 합니다.

ThirFir commented on data/src/main/java/in/koreatech/koin/data/repository/UserRepositoryImpl.kt:
```diff
@@ -167,4 +171,135 @@ class UserRepositoryImpl @Inject constructor(
+            )
+            false
+        } catch (e: HttpException) {
+            if (e.code() == 201) {
```
200번대 코드는 성공이므로 예외를 잡는 catch로 빠지지 않습니다. 구조의 수정이 필요해 보입니다. (아래 다른 코드들도 마찬가지)

jusang3057 committed:
[Fix] User API 수정

1. request에 not-null적용
2. 버전구분 : [fun]2 -> [fun]_V2
UserRepository.kt
   verifyCertificationCode: return token:String
CheckResponse.kt
   name : CheckResponse -> Duplicated
   data : BAD_REQUEST -> INVALID_FORMAT
   class : enum class -> sealed class
UserRepositoryImpl.kt
   postGeneralRegister(name, pho...) -> postGeneralRegister(token, nam...)
   postStudentRegister(name, pho...) -> postGeneralRegister(token, nam...)

jusang3057 committed:
[Fix] User API 수정

1. request에 not-null적용
2. 버전구분 : [fun]2 -> [fun]_V2
UserRepository.kt
   verifyCertificationCode: return token:String
CheckResponse.kt
   name : CheckResponse -> Duplicated
   data : BAD_REQUEST -> INVALID_FORMAT
   class : enum class -> sealed class
UserRepositoryImpl.kt
   postGeneralRegister(name, pho...) -> postGeneralRegister(token, nam...)
   postStudentRegister(name, pho...) -> postGeneralRegister(token, nam...)

jusang3057 commented:
1. request에 not-null적용
2. 버전구분 : [fun]2 -> [fun]_V2
UserRepository.kt
   verifyCertificationCode: return token:String
CheckResponse.kt
   name : CheckResponse -> Duplicated
   data : BAD_REQUEST -> INVALID_FORMAT
   class : enum class -> sealed class
UserRepositoryImpl.kt
   postGeneralRegister(name, pho...) -> postGeneralRegister(token, nam...)
   postStudentRegister(name, pho...) -> postGeneralRegister(token, nam...)

jusang3057 committed:
[Fix] User API 수정

1. request에 not-null적용

2. 버전구분 : [fun]2 -> [fun]_V2

UserRepository.kt
   verifyCertificationCode: return token:String

CheckResponse.kt
   name : CheckResponse -> Duplicated
   data : BAD_REQUEST -> INVALID_FORMAT
   class : enum class -> sealed class

UserRepositoryImpl.kt
   postGeneralRegister(name, pho...) -> postGeneralRegister(token, nam...)
   postStudentRegister(name, pho...) -> postGeneralRegister(token, nam...)

kongwoojin commented on libs/ktlint-lib-all.jar:
이 라이브러리 파일들은 무엇인가요?

kongwoojin commented on data/src/main/java/in/koreatech/koin/data/repository/UserRepositoryImpl.kt:
```diff
@@ -167,4 +171,105 @@ class UserRepositoryImpl @Inject constructor(
+    ): Boolean {
+        return try {
+            userRemoteDataSource.postStudentRegister(
+                token,
```
Token의 경우, Bearer을 앞에 붙여서 보내줘야 합니다.
presentation 레이어에서 넘겨주기 보단 data 레이어에서 API를 호출할 때 추가하는게 좋을 것 같네요

kongwoojin commented on business/src/main/java/in/koreatech/business/feature/event/writeevent/writeevent/RegisterEventNavigator.kt:
```diff
@@ -6,6 +6,7 @@ import androidx.navigation.NavType
 import androidx.navigation.compose.composable
 import androidx.navigation.compose.navigation
 import androidx.navigation.navArgument
+import `in`.koreatech.business.navigation.ADDEVENT
```
비즈니스 앱을 건드릴 필요는 없을 것 같습니다

kongwoojin reviewed (commented):
고생하셨습니다
코멘트 확인이랑 lint 수정 부탁드려요 :)

jusang3057 committed:
[Fix] User API 수정

1. Bearer 추가

jusang3057 committed:
klint관련 수정

kongwoojin reviewed (approved):
고생하셨습니다!
lint만 수정하고 머지해주세요!

jusang3057 committed:
ktlint수정

jusang3057 committed:
ktlint수정
